packages/contracts/src/dollar/libraries/LibUbiquityPool.sol

```diff
@@ -357,68 +477,106 @@ library LibUbiquityPool {
      */
     function mintDollar(
         uint256 collateralIndex,
         uint256 dollarAmount,
         uint256 dollarOutMin,
-        uint256 maxCollateralIn
+        uint256 maxCollateralIn,
+        uint256 maxGovernanceIn,
+        bool isOneToOne
     )
         internal
         collateralEnabled(collateralIndex)
-        returns (uint256 totalDollarMint, uint256 collateralNeeded)
+        returns (
+            uint256 totalDollarMint,
+            uint256 collateralNeeded,
+            uint256 governanceNeeded
+        )
     {
-        UbiquityPoolStorage storage poolStorage = ubiquityPoolStorage();
-
         require(
-            poolStorage.isMintPaused[collateralIndex] == false,
+            ubiquityPoolStorage().isMintPaused[collateralIndex] == false,
             "Minting is paused"
         );
-
         // prevent unnecessary mints
         require(
-            getDollarPriceUsd() >= poolStorage.mintPriceThreshold,
+            getDollarPriceUsd() >= ubiquityPoolStorage().mintPriceThreshold,
             "Dollar price too low"
         );
 
         // update collateral price
         updateChainLinkCollateralPrice(collateralIndex);
 
-        // get amount of collateral for minting Dollars
-        collateralNeeded = getDollarInCollateral(collateralIndex, dollarAmount);
-        require(collateralNeeded > 0, "Cannot mint with zero collateral");
+        // user forces 1-to-1 override or collateral ratio >= 100%
+        if (
+            isOneToOne ||
+            ubiquityPoolStorage().collateralRatio >=
+            UBIQUITY_POOL_PRICE_PRECISION
+        ) {
+            // get amount of collateral for minting Dollars
+            collateralNeeded = getDollarInCollateral(
+                collateralIndex,
+                dollarAmount
+            );
+            governanceNeeded = 0;
+        } else if (ubiquityPoolStorage().collateralRatio == 0) {
+            // collateral ratio is 0%, Dollar tokens can be minted by providing only Governance tokens (i.e. fully algorithmic stablecoin)
+            collateralNeeded = 0;
+            governanceNeeded = dollarAmount
+                .mul(UBIQUITY_POOL_PRICE_PRECISION)
+                .div(getGovernancePriceUsd());
+        } else {
+            // fractional, user has to provide both collateral and Governance tokens
+            uint256 dollarForCollateral = dollarAmount
+                .mul(ubiquityPoolStorage().collateralRatio)
+                .div(UBIQUITY_POOL_PRICE_PRECISION);
+            uint256 dollarForGovernance = dollarAmount.sub(dollarForCollateral);
+            collateralNeeded = getDollarInCollateral(
+                collateralIndex,
+                dollarForCollateral
+            );
+            governanceNeeded = dollarForGovernance
+                .mul(UBIQUITY_POOL_PRICE_PRECISION)
+                .div(getGovernancePriceUsd());
+        }
 
         // subtract the minting fee
         totalDollarMint = dollarAmount
```

Comment:
[HIGH] - Substracting the minting fee after computing the collateral needed will break dollar stability.

The current approach to mint is (considering the simplest scenario where CR > 100% or `isOneToOne` is set to `true`):
1. User calls `mintDollar` and specifies the amount of dollars he wants to mint via the `dollarAmount` parameter
2. The collateral needed is computed from the `dollarAmount` requested by the user
3. **After** computing the collateral needed to mint `dollarAmount`, the `mintingFee` is substracted from `dollarAmount` so that only `totalDollarMint` will be minted

This approach is wrong and breaks the peg mechanism, given that the amount of dollar minted will always be smaller than the expected by the peg algorithm. Consider the following numbers (assuming 1 dollar = 1 LUSD):
1. User wants to mint dollar and specifies a  `dollarAmount` of 1000.
2. `getDollarInCollateral` returns a `collateralNeeded` of 1000
3. After that, the fee is substracted from `dollarAmount` (let's say a 10% applies) and the final `totalDollarMint` amount to mint is 900 dollar.

The result is we have dollar always overbacked by collateral, which will draw dollar price down because every time dollar is minted the supply becomes more scarce than the amount backing it.

This situation is similar to adding fees to an ERC4626 vault and calling the `previewMint`. As [OpenZeppelin's research](https://docs.openzeppelin.com/contracts/4.x/erc4626#fees) shows, when the shares are requested, the fees are added on top of the computed assets (not substracted from them), so that this imbalance does not occur.


On the other hand, the `redeemDollar` flow is correct. When redeeming, users will specify the `dollarAmount` they want to redeem, and the first action taken will be to substract the fee (`dollarAfterFee`). Then, all computations will be performed considering the already-substracted fee, so no supply imbalance will take place.

Reply:
Original Frax contract [works](https://github.com/FraxFinance/frax-solidity/blob/967002fc7f2c5ee9e175cabc763a776a8ed03e01/src/hardhat/contracts/Frax/Pools/FraxPoolV3.sol#L390) the same way.

> This approach is wrong and breaks the peg mechanism, given that the amount of dollar minted will always be smaller than the expected by the peg algorithm

The peg mechanism is based on arbitrage opportunity (users must always be able to mint/redeem 1 Dollar token for exactly 1 USD in collateral), not on the amount of collateral.

> The result is we have dollar always overbacked by collateral, which will draw dollar price down

I don't understand why overcollateralization is bad. On the contrary, the greater the collateral/Dollar ratio the better the pool's resistance to collateral depeg.

Overall I think that overcollateralization doesn't harm the protocol and USD peg in particular.

@0xadrii Are there any similar issues on sherlock/code4arena ?

@gitcoindev @molecula451 Help

Reply:
I also always considered overcollateralization as good. Most protocols are overcollateralized by design. For LUSD (Liquity) which is also considered as current main collateral for Dollar, MCR (minimal collateral ratio) is 110%.
User deposits ETH, which is 110% value of LUSD, gets LUSD backed loan. 

https://medium.com/@lefl0wdutemp/a-deepdive-into-liquity-v2-7463beff53d4 is a good explanation.

I will have to give some more thought to @0xadrii 's findings, and read the OpenZeppelin's research as well.